rfcs/0009-defaults-and-overrides.md

```diff
@@ -0,0 +1,885 @@
+| As one who has read access to Kuadrant policies, I want to view the effective policy enforced at the traffic routed to an application, considering all active defaults and overrides at different policies(*). | Ob    | **view-effective-policy**   |
+| As a Platform Engineer, I want to view all default policy rules that may have been replaced by more specific ones(*).                                                                                          | Ob    | **view-policy-rule-status** |
+| As a Policy Manager, I want to view all gateways and/or routes whose traffic is subject to enforcement of a particular policy rule referred by name.                                                           | Ob    | **view-policy-rule-reach**  |
+
+<sup>(*) declared in the past or in the future, by myself or any other authorized user.</sup>
+
+## Writing D/O-enabled Kuadrant Policies
+
+Writing a Kuadrant policy enabled for Defaults & Overrides (D/O), to be attached to a network object, involves declaring the following fields at the first level of the spec:
+
+- `targetRef` (required): the reference to a hierarchical network object targeted by the policy, typed as a Gateway API [`PolicyTargetReference`](https://pkg.go.dev/sigs.k8s.io/gateway-api/apis/v1alpha2#PolicyTargetReference) or [`PolicyTargetReferenceWithSectionName`](https://pkg.go.dev/sigs.k8s.io/gateway-api/apis/v1alpha2#PolicyTargetReferenceWithSectionName) type
+- `defaults`: a block of _default_ policy rules with further specification of a strategy (_atomic_ set of rules or individual rules to be _merged_ into lower policies), and optional conditions for applying the defaults down through the hierarchy
+- `overrides`: a block of _override_ policy rules with further specification of a strategy (_atomic_ set of rules or individual rules to be _merged_ into lower policies), and optional conditions for applying the overrides down through the hierarchy
+- the bare policy rules block without further qualification as a default or override set of rules – e.g. the `rules` field in a Kuadrant AuthPolicy, the `limits` field in a RateLimitPolicy.
+
+Typically, one will specify either `defaults` and/or `overrides`, or just the bare set of policy rules block without further qualification as neither defaults nor overrides.
+
+In case two or more of these fields are specified, they are processed in the following order:
+1. `defaults`
+2. bare set of policy rules without qualification (treated indistinctively as another block of defaults)
+3. `overrides`
+
+Supporting specifying the bare set of policy rules at the first level of the spec, as well and simultaneously along with `defaults` and `overrides` blocks, is a strategy that aims to provide:
+1. more natural usability, especially for those who write policies attached to the lowest level of the hierarchy supported; as well as
+2. backward compatibility for policies that did not support explicit D/O and later on have moved to doing so.
```

Comment:
Thinking about this and while playing around with the potential implementation, if we are keeping the `bare set of policy rules`, it might make sense to have validation to prevent the usage of both the `bare set of policy rules` and `defaults` at the same time (i.e., mutually exclusive) 🤔 

Otherwise, I feel like this adds another layer of merging that the implementation would need to decide upon depending on the strategy used for the effective policy 🤔 

I'm somewhat undecided on whether it should be just `bare rules` or `defaults`, or keeping both. I feel `default` seems unnecessary if it acts as an intrinsic default anyway. Or are we using the keyword `default` as an indicator that the policy is an `Inherited` policy? One argument from the other side could be that having both fields can be seen as confusing also, a user can interpret that the `bare rules` will be unaffected from `Inheritrance` 🤷 Anyway, just some thoughts 🤔

Reply:
> It might make sense to have validation to prevent the usage of both the bare set of policy rules and `defaults` at the same time (i.e., mutually exclusive) 

I like that. It simplifies processing the policy – one less "layer of merging" as you said – and it gives users some guidance for them to also write policies that are easier to read and to reason about.

One who first created a policy declaring a bare set of rules (i.e. implicit defaults) and later wants to add an `overrides` block to it can easily add the `defaults` keyword on top of the existing rules, fix the indentation, and add the `overrides` part. Not a big deal. I personally prefer that than having complicated policies that mix two ways to express the same thing.

> Or are we using the keyword default as an indicator that the policy is an Inherited policy?

No, we are not. Our inherited policies are inherited policies regardless of whether the rules are explicitly grouped within D/O blocks or not. In fact, we'll label the policy CRDs as inherited so there's no doubt about it.